consensus/proto_array/src/proto_array.rs

```diff
@@ -990,6 +996,70 @@ impl ProtoArray {
             .unwrap_or(false)
     }
 
+    /// Returns `true` if `root` is equal to or a descendant of
+    /// `self.finalized_checkpoint`.
+    ///
+    /// Notably, this function is checking ancestory of the finalized
+    /// *checkpoint* not the finalized *block*.
+    pub fn is_finalized_checkpoint_or_descendant<E: EthSpec>(&self, root: Hash256) -> bool {
+        let finalized_root = self.finalized_checkpoint.root;
+        let finalized_slot = self
+            .finalized_checkpoint
+            .epoch
+            .start_slot(E::slots_per_epoch());
+
+        let mut node = if let Some(node) = self
+            .indices
+            .get(&root)
+            .and_then(|index| self.nodes.get(*index))
+        {
+            node
+        } else {
+            // An unknown root is not a finalized descendant. This line can only
+            // be reached if the user supplies a root that is not known to fork
+            // choice.
+            return false;
+        };
+
+        // The finalized and justified checkpoints represent a list of known
+        // ancestors of `node` that are likely to coincide with the store's
+        // finalized checkpoint.
+        //
+        // Don't continue checking these values for ancestors. If they don't
+        // match for the child then they're unlikely to start matching for its
+        // ancestors.
+        for checkpoint in &[
+            node.finalized_checkpoint,
+            node.justified_checkpoint,
+            node.unrealized_finalized_checkpoint,
+            node.unrealized_justified_checkpoint,
+        ] {
+            if checkpoint.map_or(false, |cp| cp == self.finalized_checkpoint) {
+                return true;
+            }
+        }
+
+        loop {
+            // If `node` is less than or equal to the finalized slot then `node`
+            // must be the finalized block.
+            if node.slot <= finalized_slot {
+                return node.root == finalized_root;
+            }
+
+            // Since `node` is from a higher slot that the finalized checkpoint,
+            // replace `node` with the parent of `node`.
+            if let Some(parent) = node.parent.and_then(|index| self.nodes.get(index)) {
+                node = parent
+            } else {
+                // If `node` is not the finalized block and its parent does not
+                // exist in fork choice, then the parent must have been pruned.
+                // Proto-array only prunes blocks prior to the finalized block,
+                // so this means the parent conflicts with finality.
+                return false;
+            };
+        }
```

Comment:
I think this `loop {}` could be written more similarly to `is_descendant` and pulled out into its own function `is_descendant_from_ancestor_at_slot`:

```rust
/// Check if the ancestor of `descendant_root` at `ancestor_slot` is `ancestor_root`.
///
/// Takes into account ancestors at skipped slots.
fn is_descendant_from_ancestor_at_slot(&self,
    ancestor_root: Hash256,
    ancestor_slot: Slot,
    descendant_root: Hash256
) -> bool {
    self.iter_block_roots(&descendant_root)
        .find(|(_root, slot)| slot <= ancestor_slot)
        .map_or(false, |(root, _)| root == ancestor_root)
}
```

Then we call it as:

```rust
self.is_descendant_from_ancestor_at_slot(finalized_root, finalized_slot, root)
```

(we _should not_ use `node.parent` as that would trigger the exact case being addressed).

Or we could inline it with similar comments? Also happy to keep the longer `loop` version, as it is more explanatory.

Reply:
The only benefit that the current approach has is that it avoids doing a double read of the first `node`. It's certainly not a big deal since it is just hashmap access + array access.

I definitely like the generalized approach and implemented it, but became a little torn when I noticed the double-read 🤔

Reply:
After coming back to this I'm tempted to leave the loop as-is. I'm very much on the fence so happy to make the change if you feel strongly ☺️

Reply:
Yeah I'm OK to leave as-is. We can refactor in future if we find we want that descendant at slot function for anything else